2.5.1

KillerOfPie opened the pull request:
[contributing]: https://github.com/SparklingComet/TradeShop/blob/master/.github/CONTRIBUTING.md

## Pull Request Etiquette

There are several guidelines you should follow in order for your Pull Request to be merged.
Replace `[ ]` with `[x]` to cross the checkbox.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Changes

Rewrite of item collection to solve bugs

KillerOfPie committed:
Rewrite of item collection to solve bugs

KillerOfPie committed:
Finished Rewrite, Multi Item Trades Tested Okay

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/ObjectHolder.java:
```diff
@@ -1,14 +1,46 @@
+		return "ObjectHolder{" +
+				"obj=" + obj +
+				'}';
+	}
```
we might as well just "serialize" this to `obj.toString()`

KillerOfPie replied:
We could, I do like being able to see that it is an object in a holder since this will almost exclusively be for debug code

SparklingComet commented on src/main/java/org/shanerx/tradeshop/objects/ShopItemStack.java:
```diff
@@ -504,6 +504,13 @@ public String serialize() {
+                "itemStack=" + getItemStack() +
+                ", shopSettings=" + shopSettings +
+                '}';
+    }
```
Is this just for printing or is it actual serialization to disk? I'm assuming the former...

KillerOfPie replied:
Just for debug printing, may also be used by the log addon since it gives a pretty good user readable output

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -476,55 +478,66 @@ public ExchangeStatus canExchangeAll(Shop shop, Inventory playerInv, int multipl
-			int count = item.getAmount() * multiplier;
+	//Returns an arraylist of the itemstacks to be removed/added, if it could not get enough of an item, will return index 0 as null and index 1 as item it could not get enough of
+	public ArrayList<ItemStack> getItems(Inventory inventory, List<ShopItemStack> search, int multiplier) {
+		Map<ItemStack, Integer> storage = new HashMap<>(), found = new HashMap<>();
```
Moving from List to Map makes a lot of sense...

KillerOfPie replied:
I'm not sure if you're being sarcastic, I don't trust using the built in amounts when it's going to be changed and compared a lot as well as being potentially severely over stacked. Also don't want differing amount getting in the way of look ups

SparklingComet replied:
I agree, I'm not being sarcastic ;)

KillerOfPie replied:
It was a reasonable thing to trigger someone depending on how it was read

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -476,55 +478,66 @@ public ExchangeStatus canExchangeAll(Shop shop, Inventory playerInv, int multipl
-			totalCount += item.getAmount() * multiplier;
-			int count = item.getAmount() * multiplier;
+	//Returns an arraylist of the itemstacks to be removed/added, if it could not get enough of an item, will return index 0 as null and index 1 as item it could not get enough of
+	public ArrayList<ItemStack> getItems(Inventory inventory, List<ShopItemStack> search, int multiplier) {
```
Can we please change the return type from `ArrayList<>` to `List<>`?

KillerOfPie replied:
Done, don't know how I left that for so long since I use List for returns most other places...

SparklingComet replied:
I agree, this is not sarcasm :)

SparklingComet replied:
I saw a couple more places, I can go through it too though

KillerOfPie replied:
If you find them and don't have time to actually deal with it then just let me know where they are

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -476,55 +478,66 @@ public ExchangeStatus canExchangeAll(Shop shop, Inventory playerInv, int multipl
 
-			debugger.log("ShopTradeListener > Item Material Being Searched for: " + item.getItemStack().getType().name(), DebugLevels.TRADE);
-			debugger.log("ShopTradeListener > Item count: " + count, DebugLevels.TRADE);
+		debugger.log("ShopTradeListener > Search List: " + search, DebugLevels.TRADE);
```
Is this actually going to pretty-print the list? Asking because I am not sure.

KillerOfPie replied:
Properly prints the list(this is why the toStrings were added) it isn't printed pretty but is completely readable outside of shulker contents(since they print as a B64)

KillerOfPie replied:
Screenshot of the output that I had put in discord
![unknown](https://user-images.githubusercontent.com/7243280/160692581-a61c0578-bb11-45df-a16f-7a51e6da83ba.png)

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -476,55 +478,66 @@ public ExchangeStatus canExchangeAll(Shop shop, Inventory playerInv, int multipl
-				if (storageItem != null) {
-					boolean isSimilar = item.isSimilar(storageItem.clone());
-					if (isSimilar) {
+		bad.add(null);
```
What's this doing?

KillerOfPie replied:
The code on the receiving end checks for `null` in the first position to determine if the return was bad or not and items in the positions(used to just check only the 1st) to tell the player what is missing. So this makes sure that if the mad list is returned the first item is always `null`

SparklingComet replied:
Hmm this is kinda hacky, not sure I like it...

KillerOfPie replied:
In the code im working on now I moved it to be part of a createBadList method so there aren't weird add(null) calls, we could change to return a Tuple<Boolean, List> but it'll make things a lot more complicated since I'm making canExchangeAll pass the result along in a Tuple already for the better messaging. Changing would end in occasionally passing around a Tuple containing a Tuple which I like less then a predictable null value check(which has been there for a bit and hasn't thrown errors.

As a note the contents of the list are either the found items or the items that could not be found.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -476,55 +478,66 @@ public ExchangeStatus canExchangeAll(Shop shop, Inventory playerInv, int multipl
+			if (itemStack != null) {
+				ItemStack tempItemStack = itemStack.clone();
+				int amount = tempItemStack.getAmount();
+				tempItemStack.setAmount(1);
```
what is this doing?

KillerOfPie replied:
Makes sure that the item-stacks used as keys all have an amount of 1. see my comment on line 483

SparklingComet reviewed (commented):
Were the contents of the list returned by `getItems()` tested?

KillerOfPie committed:
Optimization of Hopper Protection

KillerOfPie committed:
Added 1s Caching for Hopper Blocking

KillerOfPie commented:
> Were the contents of the list returned by `getItems()` tested?

I did basic testing to make sure it worked and will do more once I update the missing items messages.

KillerOfPie commented:
What do you think of adding a Guava Cache for items in DataStorage, they would save normally but we would lower the amount of disk reads we do. Also since they aren't read as ofen or quick as the Hoppers the Caches can max at like 100 and expireAfterAccess at 1-5min?

KillerOfPie committed:
Requested change from ArrayList to List

KillerOfPie committed:
Messaging and Text Display Updates + Bug Fixes
- Added ITRADE_NO_COST_AMOUNT setting and FAILED_TRADE message
- Changed INSUFFICIENT_ITEMS, SHOP_INSUFFICIENT_ITEMS, and ON_TRADE messages to show full trade details
- Message file incremented to v1.3
- Added ItemMultiLin Message method
- Changed ObjectHolder#toString to just pass Object#toString to reduce output size and readability
- DRY'd Shop#areCostsValid and Shop#areProducts Valid
- Fixed bug caused by Shop#hasProductStock and Shop#hasCostStock pulling directly from product/cost objects instead of Shop#getCost/Shop#getProduct(was causing inaccurate output from getItems the wasn't apparent until other changes were made)
- Changed Utils#getItems to take Inventory#getStorageContents rather than Inventory
- Added Static implementation of ShopItemStack#getCleanItemName to be used without creating new objects when not needed and redirected non-static method to use new static method(DRY)
- Added support to ShopItemStack#getCleanItemName for book Titles and Proper formatting of ItemType returns
- Added MISSING_ITEMS, RECEIVED_LINES, and GIVEN_ITEMS to Variables Enum
- Added Utils#createBadList to clean up creation of `null as first index lists
- Cleaned up some javaDocs and missed significantly more
- Changed return of Utils#canExchangeAll to Tuple<ExchangeStatus, List<ItemStack>> to return bad lists in cases where ExchangeStatus indicates failure do to missing items. returns badList in all other cases
- Cleaned up message calls in ShopTradeListener using new detailed messages.
- Changed ShopTradeListener#tradeAllItems to return Tuple<List<ItemStack>, List<ItemStack>> as Good/Bad CostList, ProductList for detailed messages
- Changed output for cost-free iTrade shops
- Other Cleanup and Optimizations

KillerOfPie commented:
got bored so I implemented anyways... Going to push to a separate branch so you can review and decide if we should add in or not. See #125 and leave further discussion/review there.

Below is my basic testing for it

The Red lines are the Approximate start/end of the tests

## Trade Testing
Test involves entering server, trading 5 times then returning the items for each of 10 self owned shops, then screenshot.

### Current 2.5.1 Branch:
![image](https://user-images.githubusercontent.com/7243280/160979317-4bde0792-805f-4cf8-b7ba-d76a319d30e3.png)

### With new Caching Code:
![image](https://user-images.githubusercontent.com/7243280/160979784-e5e7d469-0a49-45bf-a191-6b04158381ca.png)

## Hopper Testing
Test involves entering world and waiting for mspT(ms per Tick) to stabilize then teleporting to my hopper testing area with 80k hoppers and again waiting for mspT to stabilize. This will test if checking for shops from the storage is faster with the linkage caching.

### Current 2.5.1 Branch:
Pre-Teleport Avg mspT: ~3.3ms
Post-Teleport Avg mspT: ~78ms
Post-Teleport Peak mspT: ~150ms
Return Avg mspT: ~2.8ms

Post Test Screenshot:
Forgot to hover for peak and looks like it just GC'd, peaks looks to be around 4000mb
![image](https://user-images.githubusercontent.com/7243280/160980976-de2441ee-1e4e-46f5-bd27-0d78deaf6eb6.png)

Post Return Screenshot:
![image](https://user-images.githubusercontent.com/7243280/160981259-58c3a3cb-d361-4e83-a60e-99ae0d49ff95.png)


### With new Caching Code:
Pre-Teleport Avg mspT: ~3.1ms
Post-Teleport Avg mspT: ~ms
Post-Teleport Peak mspT: ~500ms+
Return Avg mspT: ~ms

Ended before full Peak was reached.
This probably should've been obvious since caching only helps re-accessing the items and the hoppers would have to each read from disk to check since they wouldn't be in the cache. Only option I see around this is keeping all the linkage data in memory(would have to check each world folder for linkage data on startup and load them all into memory which i'm not a fan of since it wouldn't get cleaned out of memory.

KillerOfPie committed:
Bug Fixes
- Fixes new messages not being saved properly when the message file is first created due to changes in config version 1.2
- Fixes Shop status not updating when hoppers push into a shop chest.

KillerOfPie committed:
Bug Fixes
- Fix 2 for Shop status not updating when hoppers push into a shop chest.

KillerOfPie committed:
Fix for The header in setup not being formatted

SparklingComet committed:
Update pom.xml
